[QoL] Starter UI selection update to allow removing specific pokemon from party

Opaque02 opened the pull request:
## What are the changes?
Currently when you go to add a pokemon to your party, tapping the pokemon again gives you the option to "Add to party", even if it doesn't do anything. This PR changes that behaviour so if a pokemon is in your party, you now have the option to remove in instead of adding it. It also changes the navigation of the starter selection to be more dynamic and intuitive

## Why am I doing these changes?
I got sick of going to start a run and choosing 5 pokemon only to decide my first pokemon is no good and having to remove all the pokemon until that point and then readd them all. This will fix that by letting you remove the pokemon you want no matter where they are. 

## What did change?
The starter selection ui now does a check when a pokemon is selected. If the pokemon is already in your party, the first option on the selection menu is changes the "Add to Party" text to "Remove from Party". If users choose that, it will remove the pokemon from the party at that slot and update all other arrays, including the starter icons and cursors. 

You can also navigate to the starter selection icons on the left - this will let you do all the same things that you can on the main screen (remove from party, egg moves, candies etc) without having to hunt through generations of pokemon to find the one(s) in your party. 

Next, the behaviour on live at the moment is such that if you go to choose a pokemon that's invalid, the top option will still be "Add to Party". This happens in two cases: when trying to add a pokemon in a challenge run, if the pokemon is not valid for that challenge (for example, trying to add a fire type to a monotype water challenge), the first option will be "Add to party". Secondly, if you have a pokemon that would put you over the party cost and try to add it, the first option is "Add to party". This will fail, but the option is still there. I have removed the options for "Add to party" from both of these cases.

Another thing relating to monotype challenge runs: currently in live you have to have at least one pokemon of the type as the challenge in order to add pokemon that can evolve/use an item to change into that type. For example, if you're doing a mono flying type run, you need at least one flying type pokemon added to your party before you can add pokemon that evolve into flying types. For example, even though caterpie evolves into a flying type, you need to have a flying type like pidgey in your party first. This means that the first pokemon you add will _always_ be of the same type as the challenge. If you have pidgey-caterpie and want to remove pidgey, you currently **have to** remove caterpie first. However, with this PR, since you can remove pokemon from any location, I've made some changes to make it so that if you remove all the valid pokemon from your party, all invalid pokemon (i.e. those that need to evolve to meet the type restriction) are greyed out (except your party - this was done to avoid confusion) and you won't be able to add any more invalid pokemon until you add another valid pokemon. Also if you try to start a run without any valid pokemon, you get an error message saying "This party is not valid!". You can see an example of this in the screenshots/video section.

Currently when hitting 6 pokemon in your party/hitting the cost cap before a run, the game will try to make you start the run, and if you say no, will remove the last pokemon you added. I have removed this functionality from the game so people can choose the details of the final pokemon they add to their party before hitting start.

Lastly, I have made the navigation feature more intuitive wrap around. This will be discussed in more detail later in detail.

**Things you can do:**
1. Adding pokemon to your party will be unchanged. While on a pokemon that's in the selection screen that isn't in your party, you can press Z/A/space to open up the menu. Doing so will give you the usual choices of add to party, IVs, egg moves, candy usage etc
2. Once a pokemon is in your party, selecting them again will let you choose to remove them from your party. You will still be presented with the same options as before
3. Pressing X/B/backspace is unchanged - it will remove the last pokemon you've chosen
4. In the live version, opening up the pokemon menu will always have "Add to party" as the first option, even if it's invalid. This can be seen when trying to add an invalid pokemon to your party, such as an incorrect type to a monotype run, or adding a pokemon that would put you over the part cost to your party. This feature removes the "Add to party" option for these cases, but still keeps the rest of the menu the same
5. You can now wrap around a lot more during the navigation. This will be discussed in detail below, but the short version is that you can now wrap around in more intuitive ways from the start button, generation select, starter section and party slots.

**With navigating the starter icons\*:**
1. I've added navigation from all 4 sides - scrolling from the bottom of the gen list (gen IX) will take you to the start of your party. Scrolling up from the "Start" button will take you to the bottom of your party. Scrolling left or right from the far left or far right will take you to your party sometimes - see section later on navigation logic.
2. Once you're in the party, you can go in any direction - if you're at the top, pressing up will take you to the gen list, and if you're at the bottom, going down will take you to the start button. Going left or right will take you to the pokemon select again, and put you on the same row\** as your party pokemon.
4. On the starter icons, you can press Z/A/space to open the pokemon menu again. You should see the usual options starting with remove from party. It will also let you choose your egg moves, see IVs or use candies.
7. Lastly, when you're hovering over a starter party, you'll be able to cycle through the various genders, abilities, natures, shiny, variants and form (if you've got them unlocked) without having to go back to your original choice in the starter screen.

\* one thing to note is that this section only applies if you have any pokemon in your starting party - if not, you none of this will work
\** there's some specific logic relating to which row you end up on; this is talked about later

**Starter party navigating logic:**
This expands upon the current logic for navigating between the gen select section, the pokemon section and the start button section. 

Currently, the way it works (which is how it still works if you haven't chosen any pokemon yet) is that if there's more than 5 rows of pokemon to select, anything from row 1 to row 5 (inclusive) will take you to the generation select when you go left/right from the far left/right. Anything after row 5 will take you to the start button. If there are 5 or less rows to the selection screen, _only_ the last row will take you to the start button, otherwise it'll go to the generation select. Below are examples of how this works with gen 1 (a full screen) vs gen 6 (the least amount of pokemon):

![Current and no-party navigation gen 1](https://github.com/pagefaultgames/pokerogue/assets/66582645/897f464b-bd3f-42b1-b650-4e9cc6273109)
![Current and no-party navigation gen 6](https://github.com/pagefaultgames/pokerogue/assets/66582645/d27033a5-129f-4982-9276-9136459442c6)

In the above picture, anything above the yellow line will go to the generation selector, and anything below will go to the start button.

With this update, once you have at least one pokemon in your party, the first three rows of each generation will go to the generation select, the last row of each generation will always go to the start button, and any rows in between will go to the party icons. Out of the rows that go to the party icons, the first of those rows will go to party slot 1, the second row will go to party slot 2, the third row to party slot 3 and so on. However, if you get to a row and don't have as many pokemon in your party, you will be taken to the last pokemon in your party. Inversely if there's not many rows, like with generation 6, some party slots will be skipped. Due to how this works out and the size of the starter screen, this means that only the top 5 pokemon can be navigated to.

Below is another comparison for gen 1 (a full screen) vs gen 6 (the least amount of pokemon) as per the new update:

![New party navigation gen 1](https://github.com/pagefaultgames/pokerogue/assets/66582645/2f13690e-38d4-49d1-815b-ca5908d7f8a2)
![New party navigation gen 6](https://github.com/pagefaultgames/pokerogue/assets/66582645/8f8cf6d1-7e2a-49e8-a365-35dc10a4571f)

In the above photos, anything above the top blue line will go to the generation select, anything below the yellow line goes to the start button and anything in between goes to the party icons - you can see how each row (if available) goes to a specific slot of your party icons. So for generation 1, rows 1-3 go to the generation select, rows 4-8 goes to the party icons select, and the last row (row 9) goes to the start button. Meanwhile, for generation 6, rows 1-3 go to the generation select, row 4 goes to the party select icons select and the last row (row 5) goes to the start button

**Interactions**
This final part is just to talk about some interactions with how things work (I'll add more as things go on)
1. If you go to remove the last pokemon in your party while in the starter icon section, the cursor will move up to the pokemon above the one removed. This is the case if you remove from the menu or if you have a pokemon selected in your starter icons section and you press X/B/backspace, it will remove the last pokemon you've chosen. If you remove the final pokemon in your party, you will be taken back to the generation select section. You can see this in the below video:

https://github.com/pagefaultgames/pokerogue/assets/66582645/d3e30ae8-fe8d-4b5f-b404-689046635167

### Screenshots/Videos
This is a screenshot of the starter menu before the change, even if a pokemon was already in your party:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/a3f282f2-b59c-4783-b253-486034f89912)

This is a screenshot of the menu option after the change for a pokemon that's already in your party - a pokemon not in your party still has the "Add to Party" option:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/fef857f7-cfb1-44ff-903c-49e04e350497)

Here is a video of the behaviour on the live version of the game:

https://github.com/pagefaultgames/pokerogue/assets/66582645/98c556c0-eb00-4d2f-9936-0fea9343078b

Here is a video of the behaviour after the change:

https://github.com/pagefaultgames/pokerogue/assets/66582645/ce9c17a1-4baf-4980-b6d9-75be12331a37

You'll notice that I add some pokemon to my party and then remove the weedle. Doing so removes it from the appropriate slot and moves everything up. I can then add pokemon back into my party again. From there, I move over to the far left and navigate to the starter icon section and play around with all the options

This is the behaviour in the live version of the game when going to add a pokemon that would put you over the party cost limit (you'll notice there's a bug in the current version where hovering over icons will highlight them again, even if they're invalid - this is talked about in #2386):

https://github.com/pagefaultgames/pokerogue/assets/66582645/92da8292-f4c5-4c4e-b170-f68ae99f2837

This is the behaviour for the same thing in this update (the bug in #2386 is fixed here too):

https://github.com/pagefaultgames/pokerogue/assets/66582645/f2c00870-b195-4e17-871a-fe08d4776a70

This is the behaviour for when you try to add a pokemon that doesn't meet a challenge's conditions in the live version:

https://github.com/pagefaultgames/pokerogue/assets/66582645/0243ccec-11b4-45df-ad1d-57b6abb39bc0

And this is the new behaviour in the update:

https://github.com/pagefaultgames/pokerogue/assets/66582645/22d89733-4ca7-449f-8234-df3ca90d88cd

As you can see, trying to add new pokemon to your party if they're invalid will show you the rest of the menu - toggle IVs, egg moves etc, but no more "Add to Party"

With challenges, here's a bit of a deeper dive into how it works with invalid pokemon. For monotype challenge runs, you can add a pokemon to your party that isn't the same type as the run if it evolves (or has an item to change forms) into the same type as the run; the only condition on this is that you have to already have a starter of the type of the challenge in your party before those pokemon can be added. For example, if you're doing a flying type challenge, charmander evolves into charizard, which is a flying type making it valid. However, you need to have a flying type in your party first before you can add charmander. For example, adding pidgey as your first pokemon will unlock the ability to add charmander to your party. This is how it acts in this update:

https://github.com/pagefaultgames/pokerogue/assets/66582645/2bb62a8b-705d-4755-ba13-86457e02ea86

As you can see, for this flying type run I try to add charmander and caterpie, but since only their evolutions are flying type and we don't have a flying type in our party, the "Add to Party" option isn't there. Next I go to add a pidgey to my party, which unlocks (from gen 1) magikarp, charmander and caterpie; this is because all of these pokemon evolve into a flying type pokemon eventually. I go to charmander and caterpie and add them since they're now valid. Since I now have pidgey in my party, I can start the run if I choose to. I add magikarp in - you can see that since I'm now at 9/10 cost, anything that costs more than 1 is greyed out, like the current behaviour. I take out magikarp and the sprites refresh to show what I can add given the cost is now 6, and I remove caterpie and put magikarp back in; now caterpie is not greyed out because I can still fit it into the cost of my party. 

Next I remove magikarp and pidgey from my party; when removing the latter all evolution-only pokemon that aren't in your party (in this case, magikarp), will be greyed out, and when I go to select it, I can't add it to the party. I decided to keep the pokemon in your party not greyed out so it wouldn't confuse players, though this is only as long as you have enough cost; in this case we're at 4/10 so charmander and caterpie show up, but if we were at 9/10, charmander would not show up. This is in line with how the party shows up on live at the moment. If you go to a pokemon that isn't valid but in your party and remove it (like charmander), it auto greys out once not in your party any more to let you know it's not valid any more, and you can't add it. Lastly, if you add pidgey back to your party, the evolve-only pokemon are not greyed out anymore, letting you add them back to your party. Removing pidgey once last time lets us see what the error message looks like if you go to start without a valid party.

This is how navigating to the party icons and wrap around navigation works with gen 1:

https://github.com/pagefaultgames/pokerogue/assets/66582645/f14b4c1a-974e-45ee-817c-e95b41216967

You can see me do go from the generation select to the start button by pressing the up arrow, then back to the generation by pressing down - this shows up/down wrap around. After adding some pokemon, I show how moving from the left side of the starter select screen works and which rows it lines up with. I then show what happens if we move to the left from the party icons, showing it's the same row as you'd expect, and finally show moving right/left from the start button to get to the start/end of the last row.

This is how the wrap around works with gen 6:

https://github.com/pagefaultgames/pokerogue/assets/66582645/c3019755-3251-4847-983e-cfebd2ca0af6

You can see I add pokemon to my party and show that moving from row 4 takes you to the first party slot as expected. However, since there are now less rows than pokemon slots, moving from the pokemon slots takes you to the last row. Similarly, moving left/right from the last row of pokemon starters takes you to the start button as expected

This is the logic for deciding what party slot you'll go into by moving from the starter selection section to the left. First thing to know is the row numbering system. Let's look at gen 1 as an example:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/f03914cc-f142-488e-9a85-b5c35ec22eef)

From the above, you can see the row number for each row. One thing to keep in mind is that the first three rows always go to the generation select, and the last row goes to the start button. We can use the remaining row numbers (4-8) and the table below to figure out what party slot you'll go to once you move to the party icon section. For this table, the top row shows what row you're in from the starter selection area, and the left column shows how many pokemon in your party. The intersecting number is the party slot you'll end up in (1-5 since you can't get to slot 6 from this navigation):

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/0a05fa78-c999-46e7-a5ab-a69d34c704c1)

So you can see that if you're on row 5 and you have 4 pokemon in your party, you'll go to party slot 2. This is because the row you are on is after the party slot. Similarly if you're on row 6 and have 6 pokemon in your party, you'll end up on party slot 3. 

## How to test the changes?
Download the branch and go to start a new game. Add pokemon to your party and then go to remove it using the action button (Z/A/space etc). Navigate to the starter icons section and play around with what you can do, and remove and add pokemon, and see the wraparound navigation in effect

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

travbp commented:
This would help a lot! A next evolution of this could be to make the Pokémon selectable from the party preview screen and have the pop up when they're selected there. Then, you wouldn't have to go hunt down the starter on its Gen page just to remove it or change around its moves.

Opaque02 commented:
@travbp - that's what someone in the discord mentioned as well! I've already been working on it and have got the code like 90% of the way. There's one small bug I need to fix with icons, and some logic I need to implement, and if all goes well with testing, that'll be done too :)!

travbp commented:
Sounds awesome!! Thanks for implementing it

brain-frog commented on src/ui/starter-select-ui-handler.ts:
```diff
@@ -1084,71 +1115,102 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
+          const ui = this.getUi();
+          let options = [];
+          let removeIndex = 0;
+          let isDupe = false;
```
seeing how this is being used can this be renamed to something like `isInParty`

Opaque02 replied:
This is fixed - there's now a function for isInParty that returns a [boolean, number] for isDupe and removeIndex

brain-frog commented on src/ui/starter-select-ui-handler.ts:
```diff
@@ -1084,71 +1115,102 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
+                    cursorObj.setPosition(this.cursorObj.x, this.cursorObj.y);
+                    const props = this.scene.gameData.getSpeciesDexAttrProps(species, this.dexAttrCursor);
+                    this.starterIcons[this.starterCursors.length].setTexture(species.getIconAtlasKey(props.formIndex, props.shiny, props.variant));
+                    this.starterIcons[this.starterCursors.length].setFrame(species.getIconId(props.female, props.formIndex, props.shiny, props.variant));
+                    this.checkIconId(this.starterIcons[this.starterCursors.length], species, props.female, props.formIndex, props.shiny, props.variant);
+                    this.starterGens.push(pokemonGen);
+                    this.starterCursors.push(pokemonCursor);
+                    this.starterAttr.push(this.dexAttrCursor);
+                    this.starterAbilityIndexes.push(this.abilityCursor);
+                    this.starterNatures.push(this.natureCursor as unknown as Nature);
+                    this.starterMovesets.push(this.starterMoveset.slice(0) as StarterMoveset);
+                    if (this.speciesLoaded.get(species.speciesId)) {
+                      getPokemonSpeciesForm(species.speciesId, props.formIndex).cry(this.scene);
+                    }
+                    /*
+                    if (this.starterCursors.length === 6 || this.value === this.getValueLimit()) {
+                      this.tryStart();
+                    }
+                    */
+                    this.updateInstructions();
+
+                    /**
+                    * If the user can't select a pokemon anymore,
+                    * go to start button.
+                    */
+                    /*
+                    if (!this.canAddParty) {
+                      this.startCursorObj.setVisible(true);
+                      this.setGenMode(true);
+                    }
+                    */
+
+                    ui.playSelect();
+                  } else {
+                    ui.playError(); // this should be redundant as there is now a trigger for when a pokemon can't be added to party
+                  }
+                  return true;
+                },
+                overrideSound: true
+              }];
```
please extrapolate this as a function that can be passed to `handler:` for readability

Opaque02 replied:
I've move a lot of this code out (and removed the old comments, sorry!)

There's still technically a few more things I could move out (like the if statement), but if it's not enough let me know and I can add more into the function

brain-frog commented on src/ui/starter-select-ui-handler.ts:
```diff
@@ -428,6 +432,11 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
     this.starterSelectContainer.add(this.genCursorObj);
 
+    this.starterIconsCursorObj = this.scene.add.image(this.genCursorObj.x, 64, "select_gen_cursor");
+    this.starterIconsCursorObj.setVisible(false);
```
```suggestion
    this.starterIconsCursorObj.setName("starter-icons-cursor");
    this.starterIconsCursorObj.setVisible(false);
```

Opaque02 replied:
This is fixed as per latest commit

brain-frog reviewed (changes requested):
Love these changes, I think they'll be a really nice QoL addition to the starter select

A few comments: 
If there's code that's not necessary anymore after your changes, please do not leave it commented out in the code. 
Any functions you make a significant change to please add/update the function definition comments
and with the large chaining if statements please add more comments to help with readability

Opaque02 commented:
@brain-frog - thanks for the code review and suggestions; I've gone through and removed all the old comments and added more in for nested if statements so hopefully it's easier to read, but I can add more in if not.

We've currently hit a bit of a block on this one - we were talking about it in discord not too long ago, but with monotype challenges, you can select pokemon that evolve into the type even if they aren't the type straight away. For example, if you're doing a flying monotype challenge, you can choose pidgey first which will then let you choose pokemon that evolve into a flying pokemon, like caterpie, however you can't choose caterpie until you choose pidgey. When you go to remove them, currently you remove from the back, meaning for our pidgey-caterpie starter party, we remove caterpie and are left with pidgey, which means we can add other pokemon that evolve into flying types

However with my changes, since we can now remove pokemon from any slot, there's a chance that for pidgey-caterpie, you could remove the pidgey first and then start the challenge with no current flying type pokemon. This makes the game crash since when you get to wave 1, you don't have any valid pokemon to send out. We were discussing on discord this morning ways around it and what we should do - whether we have some kind of message saying that the party is invalid, or have some other kind of thing going with it, and whether that'd be part of this PR or a new one. Until this is sorted though, I don't know if we can get this merged yet unfortunately

Opaque02 commented:
OK, that took way longer than I was expecting but all bugs are, I think, fixed now (as per my testing, hopefully). Had some weird edge cases, I think I got them all?

brain-frog commented on src/ui/starter-select-ui-handler.ts:
```diff
@@ -2010,7 +2164,7 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
       const props = this.scene.gameData.getSpeciesDexAttrProps(this.lastSpecies, dexAttr);
       const lastSpeciesIcon = (this.starterSelectGenIconContainers[this.lastSpecies.generation - 1].getAt(this.genSpecies[this.lastSpecies.generation - 1].indexOf(this.lastSpecies)) as Phaser.GameObjects.Sprite);
-      lastSpeciesIcon.setTexture(this.lastSpecies.getIconAtlasKey(props.formIndex, props.shiny, props.variant), this.lastSpecies.getIconId(props.female, props.formIndex, props.shiny, props.variant));
+      //lastSpeciesIcon.setTexture(this.lastSpecies.getIconAtlasKey(props.formIndex, props.shiny, props.variant), this.lastSpecies.getIconId(props.female, props.formIndex, props.shiny, props.variant));
```
if this line is no longer needed please delete

Opaque02 replied:
Done - thanks for the review :)
